### ISSUE TITLE
Clarification on the purpose of NewRelic::Railtie and NewRelic::Agent::Configuration::ManualSource

### ISSUE THREAD

**Issue by blowfishpro** [newrelic/newrelic-ruby-agent#1050]
## Description

[It seems like NewRelic attempts to use the Rails app configuration as a config source](https://github.com/newrelic/newrelic-ruby-agent/blob/767bb3607cfeaaae84b7181950e14fd66b6fe593/lib/newrelic_rpm.rb#L24), however I can't seem to figure out how this is supposed to work.  Tracing `init_plugin` down through the source I can see that the passed hash ends up getting set up as a `NewRelic::Agent::Configuration::ManualSource` and assigned to the `@manual_source` instance variable of NewRelic's configuration object.

But it doesn't look like there's any way to actually configure the NewRelic agent this way.  I tried setting e.g. `Rails.application.config.labels` or `Rails.application.config.newrelic_labels` but neither seem to be picked up (`NewRelic::Agent.config[:labels]` still returns `""` in the rails console)

It looks like `NewRelic::Agent.config[:config]` returns the entire rails configuration (but `config[:config]` isn't actually used by the agent).  So maybe the fact that it's `init_plugin(config: app.config)` (note the extra `config:`) is nesting the config at the wrong level.  But it also seems strange to me that the entire rails configuration would be used here unmodified.

I tried the following and it worked as expected:
```
initializer 'blah', after: 'newrelic_rpm.start_plugin' do
  newrelic_config = {}
  newrelic_config[:labels] = { foo: 'bar' }
  config.newrelic = newrelic_config
  NewRelic::Control.instance.init_plugin(newrelic_config)
end
```

In that case I can see that `NewRelic::Agent.config[:labels]` correctly returns `{ foo: 'bar' }`

## Expected Behavior

TBD.  It looks like this isn't documented anywhere so maybe it was partially implemented but never finished?  So maybe either `NewRelic::Railtie` should put `config.newrelic = {}` on the rails configuration and then use it, allowing that to be a valid method of configuring the agent, or it should not pass anything to `init_plugin`, making it explicit that no manual source is configured.

**Comment by fallwith** [comment]
Hi @blowfishpro,

Thanks very much for submitting this issue.

The New Relic Ruby agent cannot be configured via a Rails application's own configuration (`Rails.configuration`). I appreciate how confusing it must appear for us to be using a parameter named `config` and have it set equal to the Rails app's own configuration. I can readily see how that would give the appearance of that `config` possibly referencing the Ruby agent's own configuration and I can also see how with that in mind that the Rails app's configuration ends up looking like it is nested one level too deep.

The `init_plugin` method you reference in your code block (defined in [instance_methods.rb#L49](https://github.com/newrelic/newrelic-ruby-agent/blob/dev/lib/new_relic/control/instance_methods.rb#L49)) takes in an options hash as its only argument. This hash has a key of `:config` with a corresponding value of the Rails app's own configuration and gets handed off to `init_config` (which in a Rails context is defined in [rails.rb#L45](https://github.com/newrelic/newrelic-ruby-agent/blob/dev/lib/new_relic/control/frameworks/rails.rb#L45)). From there, the Rails app configuration is stored in `@config` for the express purpose of having a fallback plan in the event that a `Rails.configuration` call fails (from [rails.rb#L35](https://github.com/newrelic/newrelic-ruby-agent/blob/dev/lib/new_relic/control/frameworks/rails.rb#L35)). That call should only fail if the version of Rails being used is lower than 2.0. The current stable version of Rails is 7.0 and my guess is that you are likely using a version much newer/higher than 2.0. So ultimately the Rails app configuration originally passed to `init_plugin` is getting ignored entirely when your application successfully responds to a `Rails.configuration` call, and the `@config` variable is not even used.

The current version of the New Relic Ruby agent does not support Rails versions older/lower than v3.2, so we should likely take some time to clean up all this code that I've mentioned here. This would allow us to remove the `:config` key/value pair being passed to `init_plugin` entirely and hopefully cut down on future confusion.

The agent reads the Rails app specific configuration to glean some app specific things such as the environment value.

The New Relic Ruby agent itself should be configured via its configuration file, `newrelic.yml`. In some cases that YAML file's options can be overridden with the use of environment variables and/or server-side configuration applied by a New Relic user.

We do not currently have any plans to add additional avenues for agent configuration such as leveraging a Rails application's own configuration. If there is something you are attempting to accomplish with configuration that our existing code or documentation has yet to get you squared away with, do please let us know so that we can help.

**Comment by blowfishpro** [comment]
Alright that makes sense.  I think removing that logic would help make this a bit less ambiguous.

It would be nice to have a way to configure the agent programmatically though.  We have some pieces of configuration that we want to share across apps and right now this is done by manipulating `ENV` before the agent starts (if we did this in newrelic.yml we'd have to copy the change into dozens of apps).  We could think about setting these environment variables externally but this would be rather complicated at the moment because we're transitioning between deployment environments.

**Comment by fallwith** [comment]
Hi @blowfishpro,

We'll be using issue #1054 to track the removal of the old logic.

As for programmatic configuration, here are a few thoughts:

* By default the agent wants to try to start automatically. This behavior can be disabled by setting `agent_enabled: false` in the configuration. Then the agent can then be manually started via `NewRelic::Agent.manual_start` (see: [README](https://github.com/newrelic/newrelic-ruby-agent#other-environments), [Monitor short-lived processes](https://docs.newrelic.com/docs/apm/agents/ruby-agent/background-jobs/monitor-ruby-background-processes/#short), and the [RubyDoc](https://www.rubydoc.info/github/newrelic/newrelic-ruby-agent/NewRelic%2FAgent:manual_start)).
    * The key/value pairs contained within the optional hash argument that `manual_start` accepts will be merged with the agent's configuration. For example, the app name could be passed in like so: `NewRelic::Agent.manual_start(app_name: 'My App Name')`. Use multiple keys and values for multiple configuration parameters.
* The [Ruby agent configuration](https://docs.newrelic.com/docs/apm/agents/ruby-agent/configuration/ruby-agent-configuration/) documentation explains how the order of precedence works and what all of the options are.
* If performing a manual start won't work for your use case, then we would be happy to further explore your desired behavior via the feature request process. You (or perhaps your teammate in charge of New Relic relations) can start the process via your New Relic portal. Alternatively you can submit a new GitHub Issue to serve as a feature request.

**Comment by blowfishpro** [comment]
Okay, just a couple more questions about how that would work.

* To make that work we need to set `agent_enabled: false` somewhere (ideally one place and not in every app).  Maybe this is still easiest to do with environment variable manipulation?
* What order do things need to happen in?  Presumably the agent disabling needs to happen before the `newrelic_rpm.start_plugin` initializer.  Does the `manual_start` need to happen before or after that?

**Comment by kaylareopelle** [comment]
Hi @blowfishpro,

Thanks for your questions! I think we have a slightly different option that merges these two ideas and avoids needing to call `agent_enabled: false`.

Manually starting the agent allows you to pass almost every config option to `NewRelic::Agent.manual_start`. You shouldn't need to use `newrelic.yml` or environment variables at all.  

 First, set the installation of the `newrelic_rpm` gem in your gemfile to `require: false`: 

```
gem 'newrelic_rpm', require: false
```

This will prevent the agent from starting until `NewRelic::Agent.manual_start` is called.

 Then, in a Rails initializer, call `NewRelic::Agent.manual_start`. Your configuration options can be passed in as arguments:   
 ```
# config/initilizers/newrelic_example.rb 
require 'new_relic/agent' 

NewRelic::Agent.manual_start(
  app_name: 'Blowfishpro',
  license_key: 'XXXXXXXXXXXXXXXXX',
  enabled: true,
)
```

If you'd like to use environment variables for some configuration options, they will overwrite anything set in the `manual_start` method's arguments. The `manual_start` arguments will overwrite values in the `newrelic.yml` file.

I apologize for the pivot from your original questions and our original response. Please let me know if further clarification would be helpful!

**Comment by blowfishpro** [comment]
Alright I think I understand now, I'll try that
